cpp/src/parquet/arrow/schema.cc

```diff
@@ -681,6 +681,10 @@ Status ListToSchemaField(const GroupNode& group, LevelInfo current_levels,
       // List of primitive type
       RETURN_NOT_OK(
           NodeToSchemaField(*list_group.field(0), current_levels, ctx, out, child_field));
+    } else if (list_group.field_count() == 1 && list_group.field(0)->is_repeated()) {
```

Comment:
i'm not sure this is correct, or at least the comments above need to be updated to explain the logic further.

Specifially from the exapmles on logical type.md:

```
// List<OneTuple<String>> (nullable list, non-null elements)
optional group my_list (LIST) {
  repeated group array {
    required binary str (STRING);
  };
}
```

This seems to imply that despite how the file was written with Avro bindings there should in fact be an intermediate struct and not a llist<list<>>, its not clear to me if this is a bug in the spec, or a bug in Avro java writer implementation.

Reply:
The rule (3) of backward-compatibility rules is that `If the repeated field is a group with one field and is named either array or uses the LIST-annotated group's name with _tuple appended then the repeated type is the element type and elements are required.`. It says that **the repeated type is the element type**.

```
optional group my_list (LIST) {
  repeated group array {
    required binary str (STRING);
  };
}
```
So for the schema you've just mentioned above, its element type is `group array { required binary str (STRING); }` which perfectly resolves to `OneTuple<String>`.

```
optional group a (LIST) {
  repeated group array (LIST) {
    repeated int32 array;
  }
}
```
However, for the schema I've mentioned in this issue, its element type is `group array (LIST) { repeated int32 array; }` and it perfectly resolves to `List<String>` according to rule (1) which is `If the repeated field is not a group, then its type is the element type and elements are required.`.

The parquet-java implementation has interpreted this case in the same way: https://github.com/apache/parquet-java/blob/42cf31c0fbe4f000d4ddb1e1092c6634989ea3ca/parquet-avro/src/main/java/org/apache/parquet/avro/AvroSchemaConverter.java#L588

Reply:
The rule (4) says that `Otherwise, the repeated field's type is the element type with the repeated field's repetition.`

If I change my special schema to the one below (which does not use `array` as the element name):
```
optional group a (LIST) {
  repeated group b (LIST) {
    repeated int32 c;
  }
}
```

Then rule 4 applies to the outer list and rule 1 applies to the inner list, which results in `list<list<int32>>` again.

Reply:
I have opened https://github.com/apache/parquet-format/pull/466 to clarify things.